### PR TITLE
Rest server

### PR DESCRIPTION
**Pull request by rgudwin**
### Why was it necessary?

With the new split of the old CST into CST and cst-desktop, different kinds of visualization tools were made possible. 
The problem was that the new CST did not have a way to make the status of memories and codelets available for an external visualization tool. The current PR came to allow the exposure of all Mind states in terms of a REST service. Now, it is possible to expose Mind states for external visualization, if this is so desired. 

### How was it done?

The creation of the RESTServer class, creates the exposure of any Mind internal states for visualization, using an URL like
http://localhost:port providing a JSON version of the Mind internal states. 

### Implementation details

4 new classes were implemented and made available at br.unicamp.cst.io.rest package. Also, we created tests for the new functionalities. 

### How to test?

After creating a Mind and starting it, just create a RESTServer object, passing as parameters the Mind to be observed
and the port to serve the Mind state:

RESTServer rs = new RESTServer(mind,4011);

Then access http://localhost:4011 to have access to the Mind states. 

If you wish to have your JSON using Pretty Printing, you can use an alternative constructor:

RESTServer rs = new RESTServer(mind,4011,true);

where the third parameter is a boolean indicating a pretty printer formatter. 

A quick test was made available in TestREST test class, where a server is created, and then it is accessed and tested for correct values readings. 

### Future works

An enhancement for the current implementation might be the insertion of possibilities for having specific URL just for specific parts of the Mind. The current implementation makes available the full Mind through http.

### CODE REVIEW

**Review by rgudwin — COMMENTED** [review]
Trying to solve problems

**Review by rgudwin — COMMENTED** [review]
Trying to solve everything

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        Express app = new Express();
+        Gson gson;
+        if (pretty)
+            gson = new GsonBuilder().registerTypeAdapter(Memory.class, new InterfaceAdapter<MemoryObject>())
+                             .registerTypeAdapter(Memory.class, new InterfaceAdapter<MemoryContainer>())
+                             .setPrettyPrinting().create();
+        else 
+            gson = new Gson();
+        CorsOptions corsOptions = new CorsOptions();
+        corsOptions.setOrigin("*");
```
Comment (andre-paraense) [inline src/main/java/br/unicamp/cst/io/rest/RESTServer.java:39]: To set CORS origin as `*` can be a massive security flaw in some cases. Are we mindful of what we are doing here?

**Review by andre-paraense — CHANGES_REQUESTED** [review]
Still need to cover a bit more of the new code with tests in order o reach minimum threshold of 90% test coverage.

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java (hunk 1 above)
Comment (rgudwin) [reply src/main/java/br/unicamp/cst/io/rest/RESTServer.java:39]: Changed to allow arbitrary patterns as origins, for clients using the server.

**Review by rgudwin — COMMENTED** [review]
Made changes to attend reviewers. Basically, we changed the way CORS origin is defined. Now the user can set a pattern for origin, different from "*", even though a default of "*" can be used, if desired.

**Comment by andre-paraense** [comment]
@rgudwin 1 check is not passing because it says the new code produced in this PR is 47% covered with tests, and the threshold is 90%. Do you think we can make it to 90%?

**Comment by rgudwin** [comment]
@andre-paraense, I don't know how this metric of 47% is generated. To my understanding, the new code is 100% tested, regarding the new functionalities. I believe these automatizations are blocking the evolution of the code and becoming a bureaucratic constraint and an annoyance.

**Comment by andre-paraense** [comment]
> @andre-paraense, I don't know how this metric of 47% is generated. To my understanding, the new code is 100% tested, regarding the new functionalities. I believe these automatizations are blocking the evolution of the code and becoming a bureaucratic constraint and an annoyance.

@rgudwin the coverage is generated based on how many lines of the new code produced are "touched"  by the unit tests written. To say that 47% of the new code is covered (and therefire 53% is not), means that 53% of the new lines of code are never touched by the unit tests, hence not tested at all.

You can check which lines are not covered by clicking in the "details" next to the check, as the screenshot below shows.

Screenshot 2022-05-06 at 10 44 26
_We can clearly see here that the two files that are poorly covered are RESTServer.java and Idea.java_

Screenshot 2022-05-06 at 10 48 29
Screenshot 2022-05-06 at 10 49 29
_If you want to check which lines are not covered, you can install the CodeClimate extension (or use any other Test Coverage tool you prefer). In the screenshots above, the CodeClimate extension marks with an orange strip all code blocks that are not covered by tests. As you write unit tests that cover these lines, the coverage will grow from the current 47% towards the threshold of 90%_

The tests are there to protect the code base - in a few exceptions, they might be an exaggeration, as they are all automatic, but most of the time, like this one, they do make sense. We do not want new code not tested to be added to the code base, for instance.

**Review by rgudwin — COMMENTED** [review]
I believe all requested changes were made

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                             .setPrettyPrinting().create();
+        else 
+            gson = new Gson();
+        CorsOptions corsOptions = new CorsOptions();
+        corsOptions.setOrigin(origin);
+        app.use(Middleware.cors(corsOptions));
+        app.get("/", (req, res) -> {
+            MindJson myJson = new MindJson(m.getRawMemory().getAllMemoryObjects(),m.getCodeRack().getAllCodelets());
+            res.send(gson.toJson(myJson));
+        });
```
Comment (andre-paraense) [inline src/main/java/br/unicamp/cst/io/rest/RESTServer.java:70]: I wonder how this will affect the execution of the Codelet. These structures that are being accessed here are all meant to be synchronized. I would expect something here in the lines of what we did for the log files: everything should be put somewhere read-only, like in a queue, and picked up from there to be published through the rest server, and not directly from the Mind structures themselves. If we do it this way, the measurement will affect the execution too much, which is unintended, specially for production applications. 

My suggestion is to use the same or a very similar solution we have used for the logs.

File: src/test/java/br/unicamp/cst/io/rest/TestREST.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+import java.util.TimerTask;
+import static org.junit.Assert.assertEquals;
+import org.junit.Test;
+
+/**
+ *
+ * @author rgudwin
+ */
+public class TestREST {
+
```
Comment (andre-paraense) [inline src/test/java/br/unicamp/cst/io/rest/TestREST.java:38]: This test is not really asserting anything, it is just running the code. Would be nice if we had proper test cases, with assertions.

**Review by andre-paraense — CHANGES_REQUESTED** [review]

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java (hunk 2 above)
Comment (rgudwin) [reply src/main/java/br/unicamp/cst/io/rest/RESTServer.java:70]: @andre-paraense this case is quite different from the logs. At the logs, we were interested in saving everything, every piece of data flowing through the codelets. This requires a much more intense data flow. In this case, this is just a snapshot view of the Mind, not a constant stream. I believe that using the same kind of structures we used at the logs would cause a much higher impact on the Codelets execution. Nevertheless, a more stream-oriented version of these methods might be a possible improvement for the future.

File: src/test/java/br/unicamp/cst/io/rest/TestREST.java (hunk 3 above)
Comment (rgudwin) [reply src/test/java/br/unicamp/cst/io/rest/TestREST.java:38]: @andre-paraense, even though the test is not exhaustive, if you see in lines 225-228, we test the use of the code and assert the obtained data is the correct data flowing within the mind.

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java (hunk 2 above)
Comment (andre-paraense) [reply src/main/java/br/unicamp/cst/io/rest/RESTServer.java:70]: But what happens then if we have a client connected to this rest server that keeps polling data from the service provided periodically in milliseconds to,, for instance, render charts?

**Comment by rgudwin** [comment]
@andre-paraense, I made changes in the code to allow a refresh rate. If a second call arrives before the refresh rate is due, it returns the previous probe in the mind instead of doing a new probe within the mind.

File: src/main/java/br/unicamp/cst/io/rest/RESTServer.java (hunk 2 above)
Comment (rgudwin) [reply src/main/java/br/unicamp/cst/io/rest/RESTServer.java]: Made changes in the code to attend that.

**Comment by andre-paraense** [comment]
> @andre-paraense, I made changes in the code to allow a refresh rate. If a second call arrives before the refresh rate is due, it returns the previous probe in the mind instead of doing a new probe within the mind.

I still think we would have a problem here if, for instance, many clients would connect through the REST Server and try to plot charts of the data - because reading and writing operations would be happening at the same time in the same objects, we would likely get in a locked situation. 

I believe the best design for this task would be to have a separate "data lake" where the same data that is inside the codelets' memory objects would be repeated in different structures that could be accessed without interfering in the codelets'memory object write operations.

Nevertheless, I will approve this anyway, albeit with this caveat in mind.
